File: src/python/portfolio_management/online/error_handler.py
"""Centralized Error Handler - migrated from ERRHNDL.cbl.

Processes all online errors, logs errors to database, formats error
messages, and controls error recovery.
"""

import logging
from dataclasses import dataclass
from datetime import datetime

from portfolio_management.models.online import OnlineErrorHandling, ErrorSeverity, ErrorAction
from portfolio_management.models.common import ReturnCode

logger = logging.getLogger(__name__)

PROGRAM_ID = "ERRHNDL"


@dataclass
class ErrorLogEntry:
    timestamp: str = ""
    program: str = ""
    paragraph: str = ""
    severity: str = ""
    message: str = ""
    action_taken: str = ""


class OnlineErrorHandler:
    def __init__(self):
        self._error_log: list[ErrorLogEntry] = []
        self._error_count = 0

    def handle_error(self, error_info: OnlineErrorHandling) -> int:
        self._error_count += 1

        entry = ErrorLogEntry(
            timestamp=datetime.now().strftime("%Y-%m-%d-%H.%M.%S.%f"),
            program=error_info.program,
            paragraph=error_info.paragraph,
            severity=error_info.severity,
            message=error_info.message,
        )

        self._log_error(entry)
        formatted_msg = self._format_message(error_info)
        action = self._determine_action(error_info)
        entry.action_taken = action

        self._error_log.append(entry)

        if action == ErrorAction.ABEND:
            logger.critical("ABEND condition: %s - %s", error_info.program, formatted_msg)
            return ReturnCode.CRITICAL
        elif action == ErrorAction.RETURN:
            logger.error("Error (return): %s - %s", error_info.program, formatted_msg)
            return ReturnCode.ERROR
        else:
            logger.warning("Error (continue): %s - %s", error_info.program, formatted_msg)
            return ReturnCode.WARNING

    def _log_error(self, entry: ErrorLogEntry) -> None:
        logger.error(
            "Online Error - Program: %s, Paragraph: %s, Severity: %s - %s",
            entry.program,
            entry.paragraph,
            entry.severity,
            entry.message,
        )

    def _format_message(self, error_info: OnlineErrorHandling) -> str:
        parts = [f"Program: {error_info.program}"]

        if error_info.sqlcode != 0:
            parts.append(f"SQLCODE: {error_info.sqlcode}")
        if error_info.cics_resp != 0:
            parts.append(f"RESP: {error_info.cics_resp}")

        parts.append(f"Message: {error_info.message}")
        return " | ".join(parts)

    def _determine_action(self, error_info: OnlineErrorHandling) -> str:
        if error_info.action:
            return error_info.action

        if error_info.severity == ErrorSeverity.FATAL:
            return ErrorAction.ABEND
        elif error_info.severity == ErrorSeverity.WARNING:
            return ErrorAction.RETURN
        else:
            return ErrorAction.CONTINUE

    def get_error_log(self) -> list[ErrorLogEntry]:
        return list(self._error_log)

    @property
    def error_count(self) -> int:
        return self._error_count

File: java/src/main/java/com/cog/clbs/error/ErrorHandler.java
package com.cog.clbs.error;

import com.cog.clbs.program.ReturnCode;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.function.Consumer;

/**
 * Centralized error handler.
 *
 * <p>Java equivalent of {@code src/templates/error/error-handling.cbl}
 * (batch error handling / RC 0-16 management) and the online handler
 * {@code src/programs/online/ERRHNDL.cbl} (logging, message formatting,
 * action determination):
 *
 * <pre>
 *   8010/8020/8030-HANDLE-*  -&gt; handle(record): counts by severity
 *   8100-LOG-ERROR / P200    -&gt; error log (pluggable sink; default in-memory)
 *   P300-FORMAT-MESSAGE      -&gt; formatMessage(record)
 *   P400-DETERMINE-ACTION    -&gt; determineAction(record)
 *   8000-CHECK-FINAL-STATUS  -&gt; finalReturnCode()
 * </pre>
 */
public class ErrorHandler {

    private final List<ErrorRecord> errorLog = new ArrayList<>();
    private final Consumer<ErrorRecord> logSink;

    private int warningCount;
    private int errorCount;
    private int severeCount;
    private boolean abendRequested;

    public ErrorHandler() {
        this(null);
    }

    /** @param logSink optional external log destination (the ERRLOG INSERT equivalent) */
    public ErrorHandler(Consumer<ErrorRecord> logSink) {
        this.logSink = logSink;
    }

    /**
     * Processes one error: counts it by severity, formats the message,
     * logs it, and determines the recovery action.
     */
    public ErrorAction handle(ErrorRecord record) {
        switch (record.getSeverity()) {
            case WARNING -> warningCount++;
            case INFO -> { /* informational: no count impact */ }
            case FATAL -> severeCount++;
        }
        formatMessage(record);
        logError(record);
        return determineAction(record);
    }

    /** Records a non-fatal application error (8020-HANDLE-ERROR pattern). */
    public void recordError() {
        errorCount++;
    }

    /**
     * P300-FORMAT-MESSAGE: 'Error in {program} - {message} ({traceId})'.
     */
    public String formatMessage(ErrorRecord record) {
        String formatted = "Error in " + record.getProgram().trim()
                + " - " + record.getMessage()
                + " (" + record.getTraceId() + ")";
        record.setMessage(formatted);
        return formatted;
    }

    /** 8100-LOG-ERROR / P200-LOG-ERROR: append to the error log. */
    private void logError(ErrorRecord record) {
        errorLog.add(record);
        if (logSink != null) {
            logSink.accept(record);
        }
    }

    /**
     * P400-DETERMINE-ACTION: FATAL -&gt; ABEND, WARNING/INFO -&gt; CONTINUE,
     * otherwise RETURN.
     */
    public ErrorAction determineAction(ErrorRecord record) {
        ErrorAction action = switch (record.getSeverity()) {
            case FATAL -> ErrorAction.ABEND;
            case WARNING, INFO -> ErrorAction.CONTINUE;
        };
        if (action == ErrorAction.ABEND) {
            abendRequested = true;
        }
        record.setAction(action);
        return action;
    }

    /**
     * 8000-CHECK-FINAL-STATUS: derive the final RC 0-16 return code from
     * the accumulated severity counts (severe &gt; error &gt; warning &gt; success).
     */
    public ReturnCode finalReturnCode() {
        if (severeCount > 0) {
            return ReturnCode.SEVERE;
        }
        if (errorCount > 0) {
            return ReturnCode.ERROR;
        }
        if (warningCount > 0) {
            return ReturnCode.WARNING;
        }
        return ReturnCode.SUCCESS;
    }

    public int getWarningCount() {
        return warningCount;
    }

    public int getErrorCount() {
        return errorCount;
    }

    public int getSevereCount() {
        return severeCount;
    }

    /** WS-ABEND-FLAG: set when a fatal error requested an abend (8040 pattern). */
    public boolean isAbendRequested() {
        return abendRequested;
    }

    public List<ErrorRecord> getErrorLog() {
        return Collections.unmodifiableList(errorLog);
    }
}
